File: services/ai-analysis/src/metrics.py
"""
@file AI Analysis Prometheus 메트릭 — HTTP + Circuit Breaker + MQ 카운터
@domain ai
@layer util
@related main.py, worker.py, circuit_breaker.py

AlgoSu AI Analysis Service — Prometheus Metrics
-------------------------------------------------
규칙 근거: /docs/monitoring-log-rules.md §9

네이밍: algosu_{service}_{metric_name}_{unit}
라벨 정책: method, path(정규화), status_code — userId/traceId 금지
보안: /metrics 인증 없이 접근 가능 (클러스터 내부 전용)
"""

from __future__ import annotations

import re
import time

from fastapi import Request
from prometheus_client import (
    CONTENT_TYPE_LATEST,
    REGISTRY,
    Counter,
    Gauge,
    Histogram,
    generate_latest,
)
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
from starlette.responses import Response as StarletteResponse

# ---------------------------------------------------------------------------
# 경로 정규화 (고카디널리티 방지)
# ---------------------------------------------------------------------------
_UUID_RE = re.compile(
    r"/[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}"
)
_NUMERIC_ID_RE = re.compile(r"/\d+")
_EXCLUDED_PATHS: frozenset[str] = frozenset({"/health", "/metrics"})


def _normalize_path(path: str) -> str:
    path = _UUID_RE.sub("/:id", path)
    path = _NUMERIC_ID_RE.sub("/:id", path)
    return path


# ---------------------------------------------------------------------------
# 메트릭 정의
# ---------------------------------------------------------------------------

http_request_duration = Histogram(
    name="algosu_ai_analysis_http_request_duration_seconds",
    documentation="HTTP request duration in seconds",
    labelnames=["method", "path", "status_code"],
    buckets=(0.01, 0.05, 0.1, 0.3, 0.5, 1.0, 2.0, 5.0),
)

http_requests_total = Counter(
    name="algosu_ai_analysis_http_requests_total",
    documentation="Total number of HTTP requests",
    labelnames=["method", "path", "status_code"],
)

http_active_requests = Gauge(
    name="algosu_ai_analysis_http_active_requests",
    documentation="Number of active HTTP requests",
)

# Sprint 141 — TS(submission/github-worker) STATE_CODE와 schema 통일.
# 기존 0/0.5/1 → 0/1/2 매핑 + `name` 라벨 추가로 Grafana 대시보드 단일 panel 통합 가능.
# 단일 CB 싱글턴이지만 라벨 일관성을 위해 default name="claude-api"로 고정.
CB_NAME_CLAUDE_API = "claude-api"

circuit_breaker_state = Gauge(
    name="algosu_ai_analysis_circuit_breaker_state",
    documentation="Circuit Breaker state (0=CLOSED, 1=HALF_OPEN, 2=OPEN)",
    labelnames=["name"],
)

claude_requests_total = Counter(
    name="algosu_ai_analysis_claude_requests_total",
    documentation="Claude API call total",
    labelnames=["status"],
)

ai_quota_checks_total = Counter(
    name="algosu_ai_analysis_quota_checks_total",
    documentation="AI quota check total",
    labelnames=["result"],  # allowed, denied
)

dlq_messages_total = Counter(
    name="algosu_ai_analysis_dlq_messages_total",
    documentation="Total messages sent to DLQ",
    labelnames=["reason"],  # parse_error, process_failure
)

mq_messages_processed_total = Counter(
    name="algosu_ai_analysis_mq_messages_processed_total",
    documentation="Total MQ messages processed",
    labelnames=["result"],  # ack, nack_dlq
)

# Sprint 141 — TS schema와 통일: 0=CLOSED, 1=HALF_OPEN, 2=OPEN
# (기존 0/0.5/1 → submission/github-worker STATE_CODE와 정렬)
CIRCUIT_STATE_VALUES: dict[str, float] = {
    "CLOSED": 0.0,
    "HALF_OPEN": 1.0,
    "OPEN": 2.0,
}


def update_circuit_breaker_gauge(
    state_value: str,
    name: str = CB_NAME_CLAUDE_API,
) -> None:
    """Circuit Breaker 상태 변경 시 Gauge 업데이트.

    Sprint 141 — name 라벨 추가. 단일 CB 싱글턴은 default 사용.
    UNKNOWN state는 안전하게 CLOSED(0.0)로 매핑.
    """
    circuit_breaker_state.labels(name=name).set(
        CIRCUIT_STATE_VALUES.get(state_value, 0.0)
    )


# ---------------------------------------------------------------------------
# FastAPI 미들웨어
# ---------------------------------------------------------------------------


class PrometheusMiddleware(BaseHTTPMiddleware):
    """HTTP 요청 메트릭 수집. /health, /metrics는 제외."""

    async def dispatch(
        self,
        request: Request,
        call_next: RequestResponseEndpoint,
    ) -> StarletteResponse:
        path = request.url.path

        if path in _EXCLUDED_PATHS:
            return await call_next(request)

        method = request.method
        normalized_path = _normalize_path(path)

        http_active_requests.inc()
        start_time = time.perf_counter()
        status_code = "500"

        try:
            response = await call_next(request)
            status_code = str(response.status_code)
            return response
        except Exception:
            raise
        finally:
            duration = time.perf_counter() - start_time
            http_active_requests.dec()
            http_request_duration.labels(
                method=method,
                path=normalized_path,
                status_code=status_code,
            ).observe(duration)
            http_requests_total.labels(
                method=method,
                path=normalized_path,
                status_code=status_code,
            ).inc()


# ---------------------------------------------------------------------------
# /metrics 엔드포인트
# ---------------------------------------------------------------------------


async def metrics_endpoint() -> StarletteResponse:
    """Prometheus scraper용 /metrics. 인증 없음."""
    return StarletteResponse(
        content=generate_latest(REGISTRY),
        media_type=CONTENT_TYPE_LATEST,
    )

File: services/github-worker/src/metrics.ts
/**
 * @file GitHub Worker Prometheus 메트릭 — DLQ 카운터 + /metrics HTTP 서버
 * @domain github
 * @layer util
 * @related worker.ts, logger.ts
 *
 * 네이밍: algosu_github_worker_{metric}_{unit}
 * 보안: /metrics 인증 없이 접근 가능 (클러스터 내부 Prometheus scraper 전용)
 * 고카디널리티 라벨 금지: userId, traceId, submissionId
 */

import { createServer, IncomingMessage, ServerResponse } from 'http';
import {
  Registry,
  Counter,
  collectDefaultMetrics,
} from 'prom-client';

/**
 * 전역 prom-client Registry — CircuitBreakerManager 등 다른 모듈과 공유
 * @see circuit-breaker.ts
 */
export const registry = new Registry();
const PREFIX = 'algosu_github_worker';

collectDefaultMetrics({ register: registry, prefix: `${PREFIX}_` });

// ─── DLQ 메트릭 ────────────────────────────

/**
 * DLQ 전송 카운터 — nack(requeue=false) 시 증가
 * 라벨: reason (parse_error, process_failure)
 */
export const dlqMessagesTotal = new Counter({
  name: `${PREFIX}_dlq_messages_total`,
  help: 'Total messages sent to DLQ',
  labelNames: ['reason'] as const,
  registers: [registry],
});

/**
 * 알려진 DLQ reason 라벨 — worker.ts의 실제 inc 호출과 일치 (zero-init 대상)
 * @see worker.ts (parse_error, process_failure)
 */
export const DLQ_REASONS = ['parse_error', 'process_failure'] as const;

/**
 * DLQ 라벨 series를 0으로 사전 초기화(zero-init).
 * counter series는 .inc() 첫 호출 시 지연 생성되어 Prometheus가 값 1로 처음 관측 → DLQReceived
 * alert의 increase(...[5m])가 비교할 baseline(0) 부재로 첫 DLQ 이벤트를 놓친다(Critic R2 P2).
 * 프로세스 시작 시 0으로 초기화하면 increase가 0→1 변화를 정확히 포착한다.
 */
for (const reason of DLQ_REASONS) {
  dlqMessagesTotal.inc({ reason }, 0);
}

/**
 * MQ 메시지 처리 카운터
 * 라벨: result (ack, nack_dlq, skipped)
 */
export const mqMessagesProcessedTotal = new Counter({
  name: `${PREFIX}_mq_messages_processed_total`,
  help: 'Total MQ messages processed',
  labelNames: ['result'] as const,
  registers: [registry],
});

// ─── GitHub API Rate Limit 메트릭 ─────────────

/**
 * GitHub API Rate Limit 경고 카운터 — remaining < threshold 시 증가
 */
export const githubRateLimitWarningsTotal = new Counter({
  name: `${PREFIX}_github_rate_limit_warnings_total`,
  help: 'Total GitHub API rate limit warnings (remaining below threshold)',
  registers: [registry],
});

/**
 * GitHub API 429 (Rate Limited) 카운터
 */
export const githubRateLimitedTotal = new Counter({
  name: `${PREFIX}_github_rate_limited_total`,
  help: 'Total GitHub API 429 rate-limited responses',
  registers: [registry],
});

// ─── /metrics HTTP 서버 ─────────────────────

const METRICS_PORT = parseInt(process.env['METRICS_PORT'] ?? '9100', 10);

/**
 * /metrics HTTP 서버 시작 — Prometheus scraper 전용
 * /health 엔드포인트도 함께 제공
 */
export function startMetricsServer(): void {
  const server = createServer(async (req: IncomingMessage, res: ServerResponse) => {
    if (req.url === '/metrics' && req.method === 'GET') {
      const metrics = await registry.metrics();
      res.writeHead(200, { 'Content-Type': registry.contentType });
      res.end(metrics);
      return;
    }
    if (req.url === '/health' && req.method === 'GET') {
      res.writeHead(200, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ status: 'ok', timestamp: new Date().toISOString() }));
      return;
    }
    res.writeHead(404);
    res.end();
  });

  server.listen(METRICS_PORT);
}
